codalab/migration.py

```diff
@@ -0,0 +1,216 @@
+
+
+class Migration:
+    """
+    Base class for BundleManager tests with a CodaLab Manager uses local database.
+    ATTENTION: this class will modify real bundle database.
+    """
+
+    def __init__(self, target_store_name) -> None:
+        self.target_store_name = target_store_name
+
+    def setUp(self):
+        self.codalab_manager = CodaLabManager()
+        # self.codalab_manager.config['server']['class'] = 'SQLiteModel'
+        self.bundle_manager = BundleManager(self.codalab_manager)
+        self.download_manager = self.codalab_manager.download_manager()
+        self.upload_manager = self.codalab_manager.upload_manager()
+
+        # Create a root user
+        self.root_user_id = self.codalab_manager.root_user_id()  # root_user_id is 0.
+
+        self.target_store = self.bundle_manager._model.get_bundle_store(
+            user_id=self.root_user_id, name=self.target_store_name
+        )
+
+        self.target_store_url = self.target_store['url']
+        self.target_store_uuid = self.target_store['uuid']
+
+        assert StorageType.AZURE_BLOB_STORAGE.value == self.target_store['storage_type']
+        self.target_store_type = StorageType.AZURE_BLOB_STORAGE
+
+    def get_bundle_uuids(self, worksheet_uuid):
+        bundle_uuids = self.bundle_manager._model.get_bundle_uuids(
+            {'name': None, 'worksheet_uuid': worksheet_uuid, 'user_id': self.root_user_id},
+            max_results=None,  # return all bundles in the worksheets
+        )
+        return bundle_uuids
+
+    def is_linked_bundle(self, bundle_uuid):
+        bundle_link_url = self.bundle_manager._model.get_bundle_metadata(
+            [bundle_uuid], "link_url"
+        ).get(bundle_uuid)
+        return bundle_link_url is None
+
+    def get_bundle_location(self, bundle_uuid):
+        # bundle_locations = self.bundle_manager._model.get_bundle_locations(bundle_uuid)  # this is getting locations from
+        bundle_location = self.bundle_manager._bundle_store.get_bundle_location(bundle_uuid)
+        return bundle_location
+
+    def get_bundle(self, bundle_uuid):
+        return self.bundle_manager._model.get_bundle(bundle_uuid)
+
+    def get_bundle_info(self, bundle_uuid, bundle_location):
+        target = BundleTarget(bundle_uuid, subpath='')
+        logging.info(f"[migration] {target}")
+        try:
+            info = download_util.get_target_info(bundle_location, target, depth=0)
+            logging.info(f"[migration] {info}")
+        except Exception as e:
+            logging.info(f"[migration] Error: {str(e)}")
+            raise e
+
+        return info
+
+    def upload_to_Azure(self, bundle_uuid, bundle_location, is_dir=False):
+        # generate target bundle path
+        file_name = "contents.tar.gz" if is_dir else "contents.gz"
+        target_location = f"{self.target_store_url}/{bundle_uuid}/{file_name}"
+
+        if FileSystems.exists(target_location):
+            path_util.remove(target_location)
```

Comment:
Why remove what already exists there? Could this potentially be dangerous?

It's seem to me this would only occur if this particular bundle had data stored both on disk and in Azure blob. Is there any scenario in which this would occur? Perhaps, instead of removing, we should just log that such a bundle exists and exit without doing anything for this particular bundle?

Reply:
This could occur if the script failed in the past and we need to resume the script. So this line makes it idempotent.